core/src/main/java/io/aiven/kafka/tieredstorage/transform/RateLimitedInputStream.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2024 Aiven Oy
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package io.aiven.kafka.tieredstorage.transform;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.time.Duration;
+
+import io.github.bucket4j.Bucket;
+
+/**
+ * Input Stream with a limited rate of reading from source.
+ * Aims to provide a stable upload speed that does not affect other streams of work in a broker.
+ */
+public class RateLimitedInputStream extends InputStream {
+    private final InputStream delegated;
+    final Bucket bucket;
+
+    public RateLimitedInputStream(final InputStream delegated, final Bucket bucket) {
+        this.delegated = delegated;
+        this.bucket = bucket;
+    }
+
+    public static Bucket rateLimitBucket(final int bucketSize, final int chunkSize) {
+        if (bucketSize < chunkSize) {
+            throw new IllegalArgumentException("bucketSize must be larger or equal to chunkSize");
+        }
+        return Bucket.builder()
```

Comment:
From the bucket4j example (https://bucket4j.com/8.12.1/toc.html#concepts), I feel it would be easier to understand if you have the limit and refill set with the same value for the bucket, can both the limit and refill set to the same size (bucketSize)?  Otherwise the rate limit is really the chunkSize which might be too low.

Reply:
Good catch. I thought chunk size would be a good size, but using the same unit for capacity and refill will provide a more steady rate. Adding this suggestion.